blueocean-dashboard/src/main/js/util/smart-fetch.js

```diff
@@ -0,0 +1,211 @@
+import isoFetch from 'isomorphic-fetch';
+import dedupe from './dedupe-calls';
+
+/**
+ * How many records to fetch by default
+ */
+export const defaultPageSize = 2; // FIXME increase page size
+
+/**
+ * Freezes an object and all child properties
+ */
+const deepFreeze = (obj) => {
+    const propNames = Object.getOwnPropertyNames(obj);
+    for (let idx = 0; idx < propNames.length; idx++) {
+        const prop = obj[propNames[idx]];
+        if (prop !== null) {
+            if (typeof prop === 'object') {
+                deepFreeze(prop);
+            }
+        }
+    }
+    return Object.freeze(obj);
+};
+
+/**
+ * Validates the status is 200-299 or returns an error
+ */
+const checkStatus = (response) => {
+    if (response.status >= 300 || response.status < 200) {
+        console.error('ERROR: ', response); // eslint-disable-line no-console
+        const error = new Error(response.statusText);
+        error.response = response;
+        throw error;
+    }
+    return response;
+};
+
+/**
+ * Get the JSON response body
+ */
+const parseJSON = (rsp) => {
+    try {
+        return rsp.json();
+    } catch (err) {
+        console.error('Unable to parse JSON: ', rsp.body, err); // eslint-disable-line no-console
+        throw new Error('Invalid JSON payload', err);
+    }
+};
+
+/**
+ * Static fetch options used for every request
+ */
+const fetchOptions = { credentials: 'same-origin' };
+
+/**
+ * Mark with $success flag and freeze the object
+ */
+const successAndFreeze = obj => {
+    const out = obj;
+    out.$success = true;
+    deepFreeze(out);
+    return out;
+};
+
+/**
+ * Calls statusCallback with an object holding the current status flag
+ * either .$pending, .$success, or .$failed
+ * Prevents duplicate fetches for the same URL
+ */
+export function fetch(url, options, onData) {
+    let _onData = onData;
+    let _options = options;
+    if (!_onData && typeof _options === 'function') {
+        _onData = _options;
+        _options = null;
+    }
+    if (_onData) {
+        _onData({ $pending: true });
+        return dedupe(url, () =>
+            isoFetch(url, _options || fetchOptions) // Fetch data
+            .then(checkStatus) // Validate success
+            .then(parseJSON) // transfer to json
+            .then(successAndFreeze) // add success field & freeze graph
+            )
+            .then((data) => {
+                _onData(data);
+            })
+            .catch(err => {
+                _onData({ $failed: err });
+            });
+    }
+    // return a fake promise, a thenable
+    // so it can be resolved multiple times
+    return {
+        then: (fn) => fetch(url, _onData, fn),
+    };
```

Comment:
Yes, it's a fake promise as noted in the comment, so you can use it like:

`fetch(url, data => { ... })`

or

`fetch(url).then(data => { ... }).catch()`, which will work, as what is returned from `.then` is a real `Promise`. The issue is that the method is called multiple times, and we don't want a promise resolving and using a cached result to prevent multiple invocations of the callback...